Add hash of dao state

ManfredKarrer opened the pull request:
@sqrrm 
Here is the current state:
<img width="1198" alt="Screen Shot 2019-03-11 at 18 41 13" src="https://user-images.githubusercontent.com/1449498/54165106-51e9ba80-442d-11e9-9c79-c332b30a3364.png">

It is still missing the p2p network data monitoring and tests with reorgs.

ManfredKarrer committed:
Remove optional setting of pubKeyScript

We want to have a deterministic dao state. The PubKeyScript in the
TxOutput was the only optional field which was only set in the
dumpBlockchainData program argument was set as it was only required for
the json dump for the explorer. It has 50 bytes of data which is about
20% of the txOutput data size. We prefer atm to avoid additional
complexity which would be created if we would handle that optional
field (e.g. exclude from hash creation).
The property might be useful as well for other use cases in future.

ManfredKarrer committed:
Add comment

ManfredKarrer committed:
Merge branch 'master' into add-hash-of-dao-state

ManfredKarrer committed:
Add monitoring for hash of DaoState

ManfredKarrer committed:
Add removeLinebreaks parameter to toTruncatedString

ManfredKarrer committed:
Create chain of dao state hashes for monitoring consensus issues

We create a chain of hashes of the dao state starting from the genesis
block height and using the previous hash in the hash. This ensures that
the history need to be correct if a particular hash at a block height is
correct. We request from seed nodes the last 10 hashes and broadcast to
our peers our hash at each new block. We build our list in memory and
listen on the new onSnapShotApplied event to start building our chain
from the genesis height up to the last snapshot block and after that
from each parsed block.
If we detect a mismatch we store it in a collection and the UI can show
a warning to the user.
We added also the onDaoStateChanged handler to the DaoStateListener.
This event is called after all parsing is completed and listeners have
completed their work. We must not use time based delays in the listener
code otherwise we would get changed our dao state after that event.
To detect such we added a assert method to throw an exception if the
dao state gets changed after the allowDaoStateChange is set to false.

ManfredKarrer committed:
Add parseBlockchainComplete check

Add parseBlockchainComplete check for onParseBlockChainComplete
before calling onParseBlockChainComplete to avoid duplicated calls of
onParseBlockChainComplete. Happened if there was only one seed node.

ManfredKarrer committed:
Add comment

ManfredKarrer committed:
Add UI for Dao state monitor

Refactor dao state monitor domain.
Add support for requesting all hashed from genesis from a peer which is
in conflict (can help to find at which block the issue started).

ManfredKarrer committed:
Update UI, remove checkArgument call

ManfredKarrer committed:
Add check for address

sqrrm commented on core/src/main/java/bisq/core/dao/state/model/blockchain/BaseTxOutput.java:
```diff
@@ -45,9 +45,10 @@
 
-    // Only set if dumpBlockchainData is true
+    // Before v0.9.6 it was only set if dumpBlockchainData was set to true but we changed that with 0.9.6
+    // so that is is always set. We still need to support it because of backward compatibility.
```
Perhaps we should do a cleanup before starting testing on mainnet. That would break the regtestnet we have running now though.

ManfredKarrer commented:
Tested with DAO TESTNET and looks good so far:

<img width="2399" alt="Screen Shot 2019-03-12 at 00 33 51" src="https://user-images.githubusercontent.com/1449498/54215301-b3556c00-44b5-11e9-9bfb-db627fcca3cc.png">

ManfredKarrer committed:
Use domain data or sorting

sqrrm commented on core/src/main/java/bisq/core/dao/state/DaoStateService.java:
```diff
@@ -162,6 +173,11 @@ public int getChainHeight() {
         return daoState.getCycles();
     }
 
+    public void addCycle(Cycle firstCycle) {
```
From what I understand this is to add any cycle, not just the first one, would be better to call the argument cycle rather than firstCycle

ManfredKarrer replied:
Correct. Will change it in upcoming commits.

ManfredKarrer committed:
Show dash instead of N/A for prev hash at genesis height

ManfredKarrer committed:
Avoid logging at startup

ManfredKarrer committed:
Add snapshot support for hashChain

- Add hashChain to snapshot
- Use TreeMap instead of Map to have deterministic order
- Use eventCoordinator to make correct order of execution transparent
- Fix missing applying of nonBsqTxOutputMap at applySnapshot

ManfredKarrer committed:
Call listeners only after batch of hashes is processed

sqrrm commented on core/src/main/java/bisq/core/dao/state/monitoring/DaoStateMonitoringService.java:
```diff
@@ -0,0 +1,263 @@
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Monitors the DaoState with using a hash fo the complete daoState and make it accessible to the network for
```
```suggestion
 * Monitors the DaoState with using a hash fo the complete daoState and make it accessible to the network
```

sqrrm commented on core/src/main/java/bisq/core/dao/state/monitoring/DaoStateMonitoringService.java:
```diff
@@ -0,0 +1,263 @@
+/**
+ * Monitors the DaoState with using a hash fo the complete daoState and make it accessible to the network for
+ * so we can detect quickly if any consensus issue arise. The data does not contain any private user
+ * data so sharing it on demand has no privacy concerns.
```
Could this be an attack vector, allowing too many requests as a DOS?

ManfredKarrer replied:
On the P2P network level it would be disconnected it it gets heavy, but basically yes. We should consider to limit it somehow.

sqrrm commented on core/src/main/java/bisq/core/dao/state/monitoring/DaoStateMonitoringService.java:
```diff
@@ -0,0 +1,263 @@
+    @Override
+    public void onParseBlockChainComplete() {
+        parseBlockChainComplete = true;
+        daoStateNetworkService.addListeners();
```
Won't this add a listener every time a snapshot is reapplied. So after reorgs there would be duplicate listeners?

ManfredKarrer replied:
Good spot. Will add a flag to protect against multiple calls.

sqrrm commented on core/src/main/java/bisq/core/dao/state/monitoring/DaoStateMonitoringService.java:
```diff
@@ -0,0 +1,263 @@
+            }
+        } else {
+            // TODO check if in reorg cases it might be a valid case
+            checkArgument(height > daoStateBlockchain.getLast().getBlockHeight(),
```
Shouldn't this check be `height == daoStateBlockchain.getLast().getBlockHeight() + 1`

ManfredKarrer replied:
Yes correct. Will change and also error msg.

sqrrm commented on core/src/main/java/bisq/core/dao/state/monitoring/DaoStateMonitoringService.java:
```diff
@@ -0,0 +1,263 @@
+        if (parseBlockChainComplete) {
+            // We delay broadcast to give peers enough time to have received the block.
+            // Otherwise they would ignore our data if received block is in future to their local blockchain.
+            int delayInSec = 1 + new Random().nextInt(5);
```
Is this a reasonable assumption, that most peers get the block within 1s of each other?

ManfredKarrer replied:
Hm.... hard to  say. Prob. its too short. For testing its better ;-). Will will add a TODO to increase it later. 5-10 sec? It is not super critical fi you receive a future block you just drop it.

sqrrm commented on core/src/main/java/bisq/core/dao/state/monitoring/DaoStateMonitoringService.java:
```diff
@@ -0,0 +1,296 @@
+        daoStateBlockchain.clear();
+        daoStateNetworkService.reset();
+
+        log.info("Apply snapshot: Last daoStateHash={}", persistedDaoStateHashChain.getLast());
```
This throws an exception for an empty persistedDaoStateHashChain.

ManfredKarrer replied:
Added a check

ManfredKarrer committed:
Refactor dao state monitor

- Move dao state monitor classes to other package structure
- Add support for proposal payload monitor (WIP)
- Use generics and sub classes for common functionality

ManfredKarrer committed:
Move monitor view to new tab

ManfredKarrer committed:
Add proposal monitor view

ManfredKarrer committed:
Reflect code review comments

ManfredKarrer committed:
Refactor updateHashChain

ManfredKarrer committed:
Add num proposals

- Cleanups
- Improve logs

ManfredKarrer committed:
Add blind vote monitoring

ManfredKarrer committed:
Fix table layout issue

The setFitToRowsForTableView call at activate fixes the issue that the
views sometimes do not render when changing views.

ManfredKarrer committed:
Add peer to log

ManfredKarrer committed:
Small fixes

ManfredKarrer commented:
@sqrrm I think it is now complete. Could you cross check again? I tested re-orgs as well and seems that all works so far. Tested also backward compatibility with master and all ok.

ManfredKarrer committed:
Fix tests

ManfredKarrer committed:
Remove db store file

sqrrm reviewed (commented):
My comment on throwing an exception only happened midway through these commits. I still think it can happen but couldn't recreate it.

Everything looks good while running on the current tip. I think it's basically good to go.

ManfredKarrer committed:
Add updated db files

ManfredKarrer committed:
Change file name for DaoStateStore db file to DaoStateStore2, Cleanup

- Change onDaoStateBlockChainChanged to onChangeAfterBatchProcessing
to make it more clear that it is only called after parsing of
blockchain is completed.
- Change file name for DaoStateStore db file to DaoStateStore2
- Remove TODOs and outdated comments
- Increase delay for broadcasting

ManfredKarrer committed:
Add comments

ManfredKarrer committed:
Cleanup after code inspection

ManfredKarrer committed:
Merge branch 'master' into add-hash-of-dao-state
